contracts/modules/royalty/policies/IpPool.sol

```diff
@@ -0,0 +1,225 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.23;
+
+import { EnumerableSet } from "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+import { ReentrancyGuard } from "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+
+import { ERC20Snapshot, ERC20 } from "@openzeppelin/contracts-v4/token/ERC20/extensions/ERC20Snapshot.sol";
+import { SafeERC20 } from "@openzeppelin/contracts-v4/token/ERC20/utils/SafeERC20.sol";
+import { IERC20 } from "@openzeppelin/contracts-v4/token/ERC20/IERC20.sol";
+
+import { IRoyaltyPolicyLAP } from "../../../interfaces/modules/royalty/policies/IRoyaltyPolicyLAP.sol";
+import { IIpPool } from "../../../interfaces/modules/royalty/policies/IIpPool.sol";
+import { ArrayUtils } from "../../../lib/ArrayUtils.sol";
+import { Errors } from "../../../lib/Errors.sol";
+
+/// @title Ip Pool
+/// @notice Defines the logic for claiming royalty tokens and revenue tokens for a given IP
+contract IpPool is IIpPool, ERC20Snapshot, ReentrancyGuard {
+    using EnumerableSet for EnumerableSet.AddressSet;
+    using SafeERC20 for IERC20;
+
+    /// @notice LAP royalty policy address
+    IRoyaltyPolicyLAP public immutable ROYALTY_POLICY_LAP;
+
+    /// @notice Ip id to whom this pool belongs to
+    address public immutable IP_ID;
+
+    /// @notice Amount of unclaimed royalty tokens
+    uint32 public unclaimedRoyaltyTokens;
+
+    /// @notice Last snapshotted timestamp
+    uint256 public lastSnapshotTimestamp;
+
+    /// @notice Amount of revenue token in the ancestors pool
+    mapping(address token => uint256 amount) public ancestorsPoolAmount;
+
+    /// @notice Indicates if a given ancestor address has already claimed
+    mapping(address claimerIpId => bool) public isClaimedByAncestor;
+
+    /// @notice Amount of revenue token in the claim pool
+    mapping(address token => uint256 amount) public claimPoolAmount;
+
+    /// @notice Amount of tokens claimable at a given snapshot
+    mapping(uint256 snapshotId => mapping(address token => uint256 amount)) public claimableAtSnapshot;
+
+    /// @notice Amount of unclaimed tokens at the snapshot
+    mapping(uint256 snapshotId => uint32 tokenAmount) public unclaimedAtSnapshot;
+
+    /// @notice Indicates whether the claimer has claimed the revenue tokens at a given snapshot
+    mapping(uint256 snapshotId => mapping(address claimer => mapping(address token => bool)))
+        public isClaimedAtSnapshot;
+
+    /// @notice Royalty tokens of the pool
+    EnumerableSet.AddressSet private _tokens;
+
+    // TODO: change to beacon upgradeable contract
+    /// @notice The constructor of the IpPool
+    /// @param name The name of the pool token
+    /// @param symbol The symbol of the pool token
+    /// @param royaltyPolicyLAP The address of the royalty policy LAP
+    /// @param supply The total supply of the pool token
+    /// @param unclaimedTokens The amount of unclaimed tokens reserved for ancestors
```

Comment:
To clarify, unclaimed Royalty Tokens?

Reply:
https://github.com/storyprotocol/protocol-core-v1/pull/26/commits/07d3be2e588b110f81f5ec710051faf9d533a4e4